django/db/models/sql/query.py

```diff
@@ -1918,6 +1918,18 @@ def clear_ordering(self, force_empty):
         if force_empty:
             self.default_ordering = False
 
+    def _column_name_conflict(self, alias):
+        models = set()
+        for join in list(self.alias_map.values())[1:]:
+            model = join.join_field.related_model
+            if model in models:
+                continue
+            for field in model._meta.local_concrete_fields:
+                if field.column == alias:
+                    return True
+            models.add(model)
+        return False
```

Comment:
I don't think we need a separate hook. We can pre-calculate set of column names and reuse it for all aliases.

Reply:
Your implementation is shorter and more simple than the separate hook. but with the hook:

1. we don't process a model which is processed already.
2. we don't process extra model when we find the `alias` in the join column names.
3. we don't need to process model if `allow_aliases = False`

Reply:
>     1. we don't process a model which is processed already.

I restored this.

>     2. we don't process extra model when we find the `alias` in the join column names.

Yes, but we're doing this multiple times (for each model), so :heavy_plus_sign: / :heavy_minus_sign: . I believe it's better to pre-calculate.
 
>     3. we don't need to process model if `allow_aliases = False`
Right, we can optimize this.

Reply:
Thanks. Looks good to me.
Last point: Since we just need `column_names` in [else block](https://github.com/django/django/pull/12589/files#diff-0edd853580d56db07e4020728d59e193R1954), wouldn't it better to move the block under the else?

Reply:
In such case we will need to calculate this multiple times because it is inside a loop. Moreover `if` is for a deprecated usage, so ...